### PR TITLE
task coordinator implement messages + integrate kafka + integrate with laundry-management-service

### PR DESCRIPTION
**Pull request by AfanasievYuriy**

### CODE REVIEW

File: laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/DetailSubmissionDto.java [hunk 1]
```diff
@@ -10,5 +12,11 @@
 public class DetailSubmissionDto {
     private int detailId;
```
Comment (andrsuh) [inline laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/DetailSubmissionDto.java]: Reference types in DTOs

File: laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/DetailSubmissionDto.java [hunk 2]
```diff
@@ -10,5 +12,11 @@
...
     private int weight;
-    private long time;
+    private long duration;
+
+    public DetailSubmissionDto(OrderDetail orderDetailMessage) {
+        this.detailId = orderDetailMessage.getDetailId();
```
Comment (andrsuh) [inline laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/DetailSubmissionDto.java:22]: you can also do `this.detailId = Objects.requireNonNull(orderDetailMessage.getDetailId())`

File: laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/OrderSubmissionDto.java [hunk 3]
```diff
@@ -13,5 +15,14 @@
     private int orderId;
```
Comment (andrsuh) [inline laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/OrderSubmissionDto.java]: reference types

File: laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/DetailSubmissionDto.java (hunk 1 above)
Comment (andrsuh) [inline laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/DetailSubmissionDto.java]: validation

File: laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/OrderSubmissionDto.java [hunk 4]
```diff
@@ -13,5 +15,14 @@
...
     private List<DetailSubmissionDto> details;
     private int laundryId;
-    private int bucketId;
```
Comment (andrsuh) [inline laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/OrderSubmissionDto.java:16]: validation

File: laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/OrderSubmissionDto.java [hunk 5]
```diff
@@ -13,5 +15,14 @@
...
+    private int bucket;
+
+    public OrderSubmissionDto(OrderSubmissionEvent orderSubmissionEvent) {
```
Comment (andrsuh) [inline laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/OrderSubmissionDto.java:21]: nullable

File: laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/OrderSubmissionDto.java [hunk 6]
```diff
@@ -13,5 +15,14 @@
...
+        this.orderId = orderSubmissionEvent.getOrderId();
+        this.laundryId = orderSubmissionEvent.getLaundryId();
+        this.bucket = orderSubmissionEvent.getBucket();
+        this.details = orderSubmissionEvent.getDetailsList().stream()
```
Comment (andrsuh) [inline laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/OrderSubmissionDto.java]: npe

File: laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/kafka/consumer/TaskCoordinatorMessageConsumer.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import com.microservices.laundrymanagement.dto.OrderSubmissionDto;
+import com.microservices.laundrymanagement.service.OrderService;
+import com.microservices.taskcoordinator.api.messages.OrderSubmissionEventWrapper.OrderSubmissionEvent;
+import com.microservices.taskcoordinator.api.messages.TaskCoordinatorEventWrapper.TaskCoordinatorEvent;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.kafka.annotation.KafkaListener;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TaskCoordinatorMessageConsumer {
```
Comment (andrsuh) [inline laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/kafka/consumer/TaskCoordinatorMessageConsumer.java]: events

File: laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/kafka/consumer/TaskCoordinatorMessageConsumer.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+    private OrderService orderService;
+
+    @KafkaListener(
+            topics = "${taskcoordinator.topic.name}",
+            groupId = "${taskcoordinator.listener.name}",
+            containerFactory = "taskCoordinatorListenerContainerFactory",
+            autoStartup = "${kafka.activateConsumers}")
+    public void listen(TaskCoordinatorEvent message) {
```
Comment (andrsuh) [inline laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/kafka/consumer/TaskCoordinatorMessageConsumer.java]: event

File: laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/kafka/consumer/TaskCoordinatorMessageConsumer.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+            case ORDERSUBMISSIONEVENT: {
+                OrderSubmissionEvent event = message.getOrderSubmissionEvent();
+                System.out.println("Received OrderSubmissionEvent " + event);
+
+                orderService.submitOrder(new OrderSubmissionDto(event));
+                break;
+            }
+            default: {
+                // TODO Afanay : log and throw exception
+                throw new IllegalArgumentException("Incorrect event received");
```
Comment (andrsuh) [inline laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/kafka/consumer/TaskCoordinatorMessageConsumer.java]: just log. don't throw exception and add TODO Vlad add metric

File: laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/kafka/consumer/TaskCoordinatorMessageConsumer.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+            topics = "${taskcoordinator.topic.name}",
+            groupId = "${taskcoordinator.listener.name}",
+            containerFactory = "taskCoordinatorListenerContainerFactory",
+            autoStartup = "${kafka.activateConsumers}")
+    public void listen(TaskCoordinatorEvent message) {
+
+        switch (message.getPayloadCase()) {
+            case ORDERSUBMISSIONEVENT: {
+                OrderSubmissionEvent event = message.getOrderSubmissionEvent();
+                System.out.println("Received OrderSubmissionEvent " + event);
```
Comment (andrsuh) [inline laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/kafka/consumer/TaskCoordinatorMessageConsumer.java]: sout? use logger

File: taskcoordinator/api/build.gradle [hunk 11]
```diff
@@ -39,4 +39,6 @@ publishing {
     }
 }
 
+testClasses.finalizedBy publishToMavenLocal
```
Comment (andrsuh) [inline taskcoordinator/api/build.gradle]: remove

File: laundry-management-service/app/src/main/resources/application.properties [hunk 12]
```diff
@@ -1,17 +1,24 @@
 # Application name is used by Eureka Service Discovery, identifying traces in Zipkin
 spring.application.name=laundry-management-service
-database.schema.name=laundry_management
 server.port=8089
+
+database.schema.name=laundry_management
```
Comment (andrsuh) [inline laundry-management-service/app/src/main/resources/application.properties:5]: duplicate property documentation from TC

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/inbound/InboundLaundryStateDto.java [hunk 13]
```diff
@@ -26,4 +28,10 @@
     @NotNull
     private Integer version;
 
+    public InboundLaundryStateDto(LaundryState laundryStateMessage) {
```
Comment (andrsuh) [inline taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/inbound/InboundLaundryStateDto.java:32]: null

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderProcessedDto.java [hunk 14]
```diff
@@ -18,13 +19,19 @@
...
     @NotNull
-    InboundLaundryStateDto laundryState;
+    private InboundLaundryStateDto laundryState;
 
     @Min(value = 0L, message = "Order completion time must be positive")
     @NotNull
-    Long completionTime;
+    private Long completionTime;
+
+    public OrderProcessedDto(OrderProcessedEvent orderProcessedEvent) {
```
Comment (andrsuh) [inline taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderProcessedDto.java:33]: null

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderSubmittedDto.java [hunk 15]
```diff
@@ -22,4 +24,9 @@
     @Valid
     @NotNull
     private InboundLaundryStateDto laundryState;
+
+    public OrderSubmittedDto(OrderSubmittedEvent orderSubmittedEvent) {
```
Comment (andrsuh) [inline taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderSubmittedDto.java:29]: null

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/kafka/consumer/LaundryManagementMessageConsumer.java [hunk 16]
```diff
@@ -16,9 +20,10 @@
 import static brave.Span.Kind.CONSUMER;
 
 @Component
-// TODO sukhoa: maybe rename to something starting with "EVENT"?
-public class MessageConsumer {
-    private final Logger logger = LoggerFactory.getLogger(MessageConsumer.class);
+public class LaundryManagementMessageConsumer {
```
Comment (andrsuh) [inline taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/kafka/consumer/LaundryManagementMessageConsumer.java]: *EventConsumer

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/kafka/consumer/LaundryManagementMessageConsumerConfiguration.java [hunk 17]
```diff
@@ -17,16 +17,21 @@
 
 @EnableKafka
 @Configuration
-public class MessageConsumerConfiguration {
+public class LaundryManagementMessageConsumerConfiguration {
```
Comment (andrsuh) [inline taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/kafka/consumer/LaundryManagementMessageConsumerConfiguration.java]: EventConsumer?

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/kafka/message/LaundryManagementMessageDeserializer.java [hunk 18]
```diff
@@ -6,7 +6,7 @@
 
 import java.util.Map;
 
-public class OrderSubmittedMessageDeserializer implements Deserializer<LaundryManagementEvent> {
+public class LaundryManagementMessageDeserializer implements Deserializer<LaundryManagementEvent> {
```
Comment (andrsuh) [inline taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/kafka/message/LaundryManagementMessageDeserializer.java]: Event

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/kafka/producer/MessageSender.java [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.kafka.core.KafkaTemplate;
+import org.springframework.kafka.support.SendResult;
+import org.springframework.retry.annotation.Retryable;
+import org.springframework.stereotype.Component;
+import org.springframework.util.concurrent.ListenableFuture;
+
+import java.util.concurrent.ExecutionException;
+
+@Component
+public class MessageSender {
```
Comment (andrsuh) [inline taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/kafka/producer/MessageSender.java]: Event

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/scheduled/ScheduledEventSender.java [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Transactional;
+
+@Component
+public class ScheduledEventSender {
+
+    private EventService eventService;
+
+    @Scheduled(fixedDelay = 10000)
```
Comment (andrsuh) [inline taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/scheduled/ScheduledEventSender.java]: I think maybe 50ms?

File: laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/DetailSubmissionDto.java (hunk 2 above)
Comment (AfanasievYuriy) [reply laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/dto/DetailSubmissionDto.java:22]: cmon, we will have mappers anyway

**Review by andrsuh — APPROVED** [review]

**Pull request merged**